service/src/main/java/cz/muni/fi/pa165/soccermanager/service/MatchServiceImpl.java

```diff
@@ -53,23 +56,63 @@ public Match fetchById(long matchId) {
     }
 
     @Override
-    public Match createMatch(Match match) {
-        matchDao.insert(match);
-        return match;
+    public Match createMatch(Match match) throws SoccerManagerServiceException {
+
+        if (match != null) {
+            if (matchDao.fetchAll().contains(match)) {
+                throw new SoccerManagerServiceException(
+                        "Match " + match.getHomeTeam().getName() +
+                                " vs. " + match.getAwayTeam().getName() + " already exists");
+            }
+
+            Team homeTeam = match.getHomeTeam();
+            Team awayTeam = match.getAwayTeam();
+            LocalDate date = match.getDate();
+
+            if (homeTeam.equals(awayTeam)) {
+                throw new SoccerManagerServiceException(
+                        "It is not allowed to create match between two same teams "
+                                + homeTeam.getName() + ". ");
+            } else if (date.isBefore(LocalDate.now())) {
+                throw new SoccerManagerServiceException(
+                        "It is not allowed to create match in the past "
+                                + date + ". ");
+            } else {
+                matchDao.insert(match);
+                return match;
+            }
+        }
+        else
+            throw new IllegalArgumentException();
     }
 
     @Override
-    public boolean isFinished(Match match) {
-        return match.isFinished();
+    public boolean isFinished(Match match) throws SoccerManagerServiceException {
+
+        if (!matchDao.fetchAll().contains(match)) {
+            throw new SoccerManagerServiceException(
+                    "Match " + match.getHomeTeam().getName() +
+                            " vs. " + match.getAwayTeam().getName() + " do not exists.");
+        }
+
+        return matchDao.isFinished(match);
     }
 
     @Override
-    public void updateMatch(Match match) {
+    public void updateMatch(Match match) throws SoccerManagerServiceException {
+
+        if (!matchDao.fetchAll().contains(match)) {
+            throw new SoccerManagerServiceException(
+                    "Match " + match.getHomeTeam().getName() +
+                    " vs. " + match.getAwayTeam().getName() + " do not exists.");
+        }
+
         matchDao.update(match);
```

Comment:
I think this is not what you want to do - If you are updating match you are already passing changed object. Contains compares objects based on equals() method so it only compares chosen attributes. If you change some attribute compared in equals then it will fail. I think we don't need this check here.